File: src/player.zig
const std = @import("std");
const rl = @import("raylib/raylib.zig");

const config = @import("config.zig");

const Flame = @import("particles/flame.zig").Flame;

const print = std.debug.print;
const math = std.math;

fn get_angle(diff: rl.Vector2) rl.Vector2 {
    var angle: f32 = math.atan2(f32, diff.x, diff.y);
    if (angle < 0) {
        angle += math.pi * 2.0;
    }
    return rl.Vector2{ .x = math.sin(angle), .y = math.cos(angle) };
}

fn get_direction(x: f32, y: f32) rl.Vector2 {
    var dir = rl.Vector2{ .x = 0.0, .y = 0.0 };
    for (config.key_right) |key, i| {
        _ = i;
        if (rl.IsKeyDown(key)) {
            dir.x = 1;
        }
    }
    for (config.key_left) |key, i| {
        _ = i;
        if (rl.IsKeyDown(key)) {
            dir.x = -1;
        }
    }
    for (config.key_up) |key, i| {
        _ = i;
        if (rl.IsKeyDown(key)) {
            dir.y = -1;
        }
    }
    for (config.key_down) |key, i| {
        _ = i;
        if (rl.IsKeyDown(key)) {
            dir.y = 1;
        }
    }
    if (dir.y == 0 and dir.x == 0) {
        // check mouse press
        if (rl.IsMouseButtonDown(rl.MouseButton.MOUSE_BUTTON_LEFT)) {
            var diff = rl.Vector2{ .x = config.mouse_x - x + config.cx, .y = config.mouse_y - y };
            const offset = 5;
            if (@fabs(diff.x) > offset or @fabs(diff.y) > offset) {
                dir = get_angle(diff);
            }
        }
    }
    return dir;
}
pub const Player = struct {
    flame: Flame = Flame{},
    position: rl.Vector2 = undefined,
    hp: f32 = 100,
    xp: f32 = 0,
    speed: f32 = 0.5,
    frozen: bool = false,

    pub fn init(self: *Player, allocator: std.mem.Allocator) void {
        self.position = rl.Vector2{ .x = config.cx + config.screen_width * 0.5, .y = config.screen_height * 0.8 };
        self.flame.init(allocator);
    }
    pub fn get_radius(self: *Player) f32 {
        return self.flame.radius * self.flame.scale;
    }
    pub fn update(self: *Player) !void {
        const x = self.position.x;
        const y = self.position.y;
        var dir: rl.Vector2 = rl.Vector2{ .x = 0, .y = 0 };
        if (!self.frozen) {
            dir = get_direction(x, y);
        }
        var dx = dir.x * self.speed * self.get_radius();
        var dy = dir.y * self.speed * self.get_radius();
        var eye_bound = config.screen_width / 5;
        if ((x + dx < config.cx + eye_bound and config.cx > 0) or (x + dx > config.cx + config.screen_width - eye_bound and config.cx < config.end_x - config.screen_width)) {
            config.cx += dx;
        }

        if (x + dx < config.cx + self.get_radius() and dx < 0) {
            self.position.x = config.cx + self.get_radius();
        } else if (x + dx > config.cx + config.screen_width - self.get_radius()) {
            self.position.x = config.cx + config.screen_width - self.get_radius();
        } else {
            self.position.x += dx;
        }
        // y limits
        var min_y = config.start_y - self.get_radius() * 0.5;
        var max_y = config.screen_height - self.get_radius();
        if (y + dy > max_y and dy > 0) {
            self.position.y = max_y;
        } else if (y + dy < min_y and dy < 0) {
            self.position.y = min_y;
        } else {
            self.position.y += dy;
        }
        try self.flame.update(self.position);
    }
    pub fn draw(self: *Player) void {
        self.flame.draw();
    }
    pub fn deinit(self: *Player) void {
        self.flame.deinit();
    }
};

File: src/player.nim
import raylib, std/random, utils, std/math, state

const
  FIREBALL_SHADER_PATH = "resources/shaders/glsl330/fireball.fs"
  FLAME_SHADER_PATH = "resources/shaders/glsl330/flame.fs"

type
  PlayerState* = enum
    Grounded = 0, Jumping, Falling, Frozen

  Player* = ref object of RootObj
    rotation: float = 0.0
    position*, velocity*: Vector3
    xp*: float = 0.0
    speed: float = 30.0
    radius*: float = 20.0
    lifetime: float = 30
    lastDirection: float = 1.0
    jumpHeight: float = 200.0
    state*: PlayerState = Grounded
    bd*: Vector3 = Vector3()
    fireball_shader: Shader
    flame_shader: Shader
    seconds: float = 0.0
    bufferAShader: Shader
    framebuffer: RenderTexture2D



const
  KEY_RIGHT = [Right, KeyboardKey(D)]
  KEY_LEFT = [Left, KeyboardKey(A)]
  KEY_UP = [Up, KeyboardKey(W)]
  KEY_DOWN = [Down, KeyboardKey(S)]

proc getDirection(self: Player): Vector3 =
  result = Vector3()
  for key in KEY_RIGHT:
    if isKeyDown(key): result.x = 1
  for key in KEY_LEFT:
    if isKeyDown(key): result.x = -1
  for key in KEY_UP:
    if isKeyDown(key): result.z = -1
  for key in KEY_DOWN:
    if isKeyDown(key): result.z = 1
  
  if result.z == 0 and result.x == 0 and isMouseButtonDown(Left):
    let mouseRay = getScreenToWorldRay(gMousePosition, gCamera)
    let collision = getRayCollisionBox(mouseRay, getBoundingBox(self.position, self.radius))
    result.x = collision.normal.x
    result.z = -(collision.normal.y + 0.5) + (collision.normal.z + 0.5)

proc getVelocity*(self: Player, dt: float): Vector3 =
    var vel = self.velocity
    let dir = self.getDirection()
    vel.x = (dir.x * self.speed * self.radius) * dt
    vel.z = (dir.z * self.speed * self.radius) * dt
    if self.state != Grounded:
      vel.y -= gravity * dt
    if self.state == Grounded and isKeyDown(Space):
      vel.y = self.jumpHeight * dt
      self.state = Jumping
  
    result = vel
  
method init*(self: Player) {.base.} =
  randomize()
  self.position = Vector3(x: gGroundSize.x * 0.4, y: gGroundSize.y + self.radius * 2, z: gGroundSize.z * 0.5)
  # Load the flame effect shader
  self.fireball_shader = loadShader("", FIREBALL_SHADER_PATH)
  self.flame_shader = loadShader("", FLAME_SHADER_PATH)
  
  if self.fireball_shader.id == 0:
    echo "Error: Failed to load flame effect shader!"

  
method update*(self: Player, dt: float) {.base.} =
  self.seconds += dt

  if self.state != Frozen:
    let
      diameter = self.radius * 2
      vel = self.velocity
      pos = self.position
    
    if (pos.x + vel.x - diameter > 0 or vel.x > 0) and (pos.x + vel.x + diameter < gGroundSize.x or vel.x < 0):
      self.position.x += vel.x
    if (pos.z + vel.z - diameter > 0 or vel.z > 0) and (pos.z + vel.z + diameter < gGroundSize.z - tileSize * 0.5 or vel.z < 0):
      self.position.z += vel.z
    self.position.y += vel.y

  let iTimeLoc = getShaderLocation(self.fireball_shader, "iTime")
  let iResolutionLoc = getShaderLocation(self.fireball_shader, "iResolution")
  let iMouseLoc = getShaderLocation(self.fireball_shader, "iMouse")
  setShaderValue(self.fireball_shader, iTimeLoc, self.seconds.float32)
  setShaderValue(self.fireball_shader, iResolutionLoc, [screenWidth.float32, screenHeight.float32])
  let mousePos = getMousePosition()
  let mouseState = Vector4(
      x: mousePos.x.float32,
      y: mousePos.y.float32,
      z: if isMouseButtonDown(MouseButton.LEFT): 1.0 else: 0.0,
      w: if isMouseButtonDown(MouseButton.RIGHT): 1.0 else: 0.0
  )
  setShaderValue(self.fireball_shader, iMouseLoc, mouseState)

method draw*(self: Player) {.base.} =
  beginBlendMode(BlendMode.Alpha);
  beginShaderMode(self.fireball_shader)

  drawSphere(self.position, self.radius * 2, RAYWHITE)

  #drawCylinder(bottomPosition, self.radius, self.radius, capsuleHeight, 16, RAYWHITE)

  endShaderMode()
  endBlendMode()
